rllib/core/optim/rl_optimizer.py

```diff
@@ -0,0 +1,72 @@
+import abc
+from typing import Mapping, Any
+
+from ray.rllib.core.rl_module import RLModule
+from ray.rllib.utils.annotations import (
+    OverrideToImplementCustomLogic,
+    OverrideToImplementCustomLogic_CallToSuperRecommended,
+    PublicAPI,
+)
+
+
+@PublicAPI(stability="beta")
+class RLOptimizer(abc.ABC):
+    """Base class for defining a loss function and optimizer for a RLModule.
+
+    Args:
+        rl_module: The RLModule that will be optimized.
+        config: The configuration for the optimizer.
+
+    Abstract Methods:
+        compute_loss: computing a loss to optimize rl_module over.
+        construct_optimizers: constructing the optimizers for rl_module.
+
+    Example:
+    .. code-block:: python
+
+        module = RLModule(...)
+        rl_optim = RLOptimizer(module, config)
+        optimizers = rl_optim.construct_optimizers()
+        sample_batch = ...
+        fwd_out = module.forward_train(sample_batch)
+        loss_dict = rl_optim.compute_loss(fwd_out)
+
+        # compute gradients of loss w.r.t. trainable variables
+        ...
+
+        for optim in optimizers:
+            optim.step()
+            optim.zero_grad()
+
+    """
+
+    @OverrideToImplementCustomLogic_CallToSuperRecommended
+    def __init__(self, module: RLModule, config: Mapping[str, Any]):
+        self.module = module
+        self._config = config
+
+    @abc.abstractmethod
+    def compute_loss(self, fwd_out: Mapping[str, Any]) -> Mapping[str, Any]:
```

Comment:
Maybe this is a little thinking too ahead: We should define whether `fwd_out` may be nested dicts as well. Might have an effect on e.g. making JAX/torch tracing of the loss function easier, but making some use cases harder (where a nested dict makes sense due to maybe a lot of loss terms that need to be structured).

Reply:
I wan't to avoid this for now. I need to test it out with tensorflow 2 first. Ideally we make this api in such a way to avoid the retracing operation as much as possible.

Reply:
see: https://www.tensorflow.org/guide/function#rules_of_tracing

Reply:
fwd_out should be a nested dict which we can add later. What we have right now in this PR is fine (Mapping[str, Any]), we can specify the exact type hint as we make more progress with this transition. In this PR, it's very important to get the compute_loss args right. Basically the user will get stuck with the parameters that we provide here. Other than fwd_out I think we also need the sample_batch (or let's call it `fwd_in`). e.g. Think of MSE loss that needs the ground truth to get computed.